cesium/internal/domain/delete.go

```diff
@@ -11,235 +11,355 @@ package domain
-			TimeRange: telem.TimeRange{
-				Start: start.Start,
-				End:   tr.Start,
-			},
-			fileKey: start.fileKey,
-			offset:  start.offset,
-			length:  uint32(startOffset), // length from start.Start to tr.Start
+			TimeRange: telem.TimeRange{Start: start.Start, End: tr.Start},
+			fileKey:   start.fileKey,
+			offset:    start.offset,
+			length:    uint32(startOffset), // length from start.Start to tr.Start
 		})
 	}
 
-	if uint32(endOffset) != end.length {
+	if endOffset != 0 {
 		newPointers = append(newPointers, pointer{
-			TimeRange: telem.TimeRange{
-				Start: tr.End,
-				End:   end.End,
-			},
-			fileKey: end.fileKey,
-			offset:  end.offset + uint32(endOffset),
-			length:  end.length - uint32(endOffset), // length from tr.End to end.End
+			TimeRange: telem.TimeRange{Start: tr.End, End: end.End},
+			fileKey:   end.fileKey,
+			offset:    end.offset + end.length - uint32(endOffset),
+			length:    uint32(endOffset), // length from tr.End to end.End
 		})
 	}
 
-	temp := make([]pointer, len(db.idx.mu.pointers)+len(newPointers)-1)
-	copy(temp, db.idx.mu.pointers[:startPosition])
-
 	if len(newPointers) != 0 {
-		copy(temp[startPosition:startPosition+len(newPointers)], newPointers)
+		db.idx.mu.pointers = append(
+			db.idx.mu.pointers[:startPosition],
+			append(newPointers, db.idx.mu.pointers[startPosition:]...)...,
+		)
 	}
 
-	copy(temp[startPosition+len(newPointers):], db.idx.mu.pointers[startPosition+1:])
-
-	db.idx.mu.pointers = temp
-
-	return nil
+	persist := db.idx.indexPersist.prepare(startPosition)
+	// We choose to keep the mutex locked while persisting to index.
+	return span.Error(persist())
 }
 
-func (db *DB) CollectTombstones(ctx context.Context, maxSizeRead uint32) error {
-	db.idx.mu.Lock()
-	defer db.idx.mu.Unlock()
+// GarbageCollect rewrites all files that are over the size limit of a file and has
+// enough tombstones to garbage collect, as defined by GCThreshold.
+func (db *DB) GarbageCollect(ctx context.Context) error {
+	ctx, span := db.T.Bench(ctx, "garbage_collect")
+	defer span.End()
 
-	for fileKey, tombstones := range db.idx.mu.tombstones {
-		var (
-			// priorTombstoneLength tracks the number of lengths of all tombstone
-			// pointers before the current pointer during the course of rewriting
-			// pointers.
-			priorTombstoneLength uint32 = 0
-			pointerPtr                  = 0
-			tombstonePtr                = 0
-		)
+	_, err := db.files.gcWriters()
+	if err != nil {
+		return span.Error(err)
+	}
 
-		r, err := db.files.acquireReader(ctx, fileKey)
-		if err != nil {
-			return err
+	for fileKey := uint16(1); fileKey <= uint16(db.files.counter.Value()); fileKey++ {
+		if db.files.hasWriter(fileKey) {
+			continue
 		}
-		f, err := db.FS.Open(strconv.Itoa(int(fileKey))+"_temp.domain", os.O_CREATE|os.O_RDWR)
+		s, err := db.FS.Stat(fileKeyToName(fileKey))
 		if err != nil {
-			return err
+			return span.Error(err)
+		}
+		if s.Size() < int64(db.FileSize) {
```

Comment:
This is a bad check. What if the file is 4 bytes underneath the max. Will we still open a new writer on it? I guess so, but we probably shouldn't because we'll end up with a tiny domain. Probably a good idea to add a better check than this.